internal/controller/apiextensions/usage/reconciler.go

```diff
@@ -0,0 +1,405 @@
+	apiextensionscontroller "github.com/crossplane/crossplane/internal/controller/apiextensions/controller"
+	"github.com/crossplane/crossplane/internal/usage"
+)
+
+const (
+	timeout   = 2 * time.Minute
+	finalizer = "usage.apiextensions.crossplane.io"
+	// Note(turkenh): In-use label enables the "DELETE" requests on resources
+	// with this label to be intercepted by the webhook and rejected if the
+	// resource is in use.
+	inUseLabelKey        = "crossplane.io/in-use"
+	detailsAnnotationKey = "crossplane.io/usage-details"
+
+	errGetUsage             = "cannot get usage"
+	errResolveSelectors     = "cannot resolve selectors"
+	errListUsages           = "cannot list usages"
+	errGetUsing             = "cannot get using"
+	errGetUsed              = "cannot get used"
+	errAddOwnerToUsage      = "cannot update usage resource with owner ref"
+	errAddDetailsAnnotation = "cannot update usage resource with details annotation"
+	errAddInUseLabel        = "cannot add in use use label to the used resource"
+	errRemoveInUseLabel     = "cannot remove in use label from the used resource"
+	errAddFinalizer         = "cannot add finalizer"
+	errRemoveFinalizer      = "cannot remove finalizer"
+	errUpdateStatus         = "cannot update status of usage"
+)
+
+// Event reasons.
+const (
+	reasonResolveSelectors event.Reason = "ResolveSelectors"
+	reasonListUsages       event.Reason = "ListUsages"
+	reasonGetUsed          event.Reason = "GetUsedResource"
+	reasonGetUsing         event.Reason = "GetUsingResource"
+	reasonDetailsToUsage   event.Reason = "AddDetailsToUsage"
+	reasonOwnerRefToUsage  event.Reason = "AddOwnerRefToUsage"
+	reasonAddInUseLabel    event.Reason = "AddInUseLabel"
+	reasonRemoveInUseLabel event.Reason = "RemoveInUseLabel"
+	reasonAddFinalizer     event.Reason = "AddFinalizer"
+	reasonRemoveFinalizer  event.Reason = "RemoveFinalizer"
+
+	reasonUsageConfigured event.Reason = "UsageConfigured"
+	reasonWaitUsing       event.Reason = "WaitingUsingDeleted"
+)
+
+type selectorResolver interface {
+	resolveSelectors(ctx context.Context, u *v1alpha1.Usage) error
+}
+
+// Setup adds a controller that reconciles Usages by
+// defining a composite resource and starting a controller to reconcile it.
+func Setup(mgr ctrl.Manager, o apiextensionscontroller.Options) error {
+	name := "usage/" + strings.ToLower(v1alpha1.UsageGroupKind)
+	r := NewReconciler(mgr,
+		WithLogger(o.Logger.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Usage{}).
+		WithOptions(o.ForControllerRuntime()).
+		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
+}
+
+// ReconcilerOption is used to configure the Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
```

Comment:
This applies to `Reconciler`, `NewReconciler` etc. and I agree that they could be private since on external usage.

However, I prefer to stay consistent with the rest of the codebase here.